imports/client/components/CallSection.tsx

```diff
@@ -161,79 +223,552 @@ const RTCCallSection = (props: RTCCallSectionProps) => {
+  // the connection parameters have been passed to the server-side transport.
+  // Therefore, stash the acknowledgement callback on a ref and call it once the
+  // corresponding ConnectAck db record is created.
+  const [sendTransport, sendTransportConnectCallback] = useTransport(device, sendServerParams);
+  const [recvTransport, recvTransportConnectCallback] = useTransport(device, recvServerParams);
+  useEffect(() => {
+    const observer = ConnectAcks.find({ peer: selfPeer._id }).observeChanges({
+      added: (_id, fields) => {
+        if (fields.direction === 'send') {
+          sendTransportConnectCallback.current?.();
+          sendTransportConnectCallback.current = undefined;
+        } else if (fields.direction === 'recv') {
+          recvTransportConnectCallback.current?.();
+          recvTransportConnectCallback.current = undefined;
+        }
+      },
+    });
+    return () => observer.stop();
+  }, [selfPeer._id, sendTransportConnectCallback, recvTransportConnectCallback]);
+
+  if (!sendTransport || !recvTransport) {
+    // No JoiningCall warning here - if we have params coming in we're
+    // guaranteed to create the local transports
+    return null;
   }
 
-  const callerCount = tracker.participants.length;
-  const others = nonSelfParticipants();
+  return (
+    <Callers
+      puzzleId={puzzleId}
+      muted={muted}
+      deafened={deafened}
+      audioContext={audioContext}
+      localStream={localStream}
+      callersExpanded={callersExpanded}
+      onToggleCallersExpanded={onToggleCallersExpanded}
+      otherPeers={otherPeers}
+      sendTransport={sendTransport}
+      recvTransport={recvTransport}
+    />
+  );
+};
 
-  const callersHeaderIcon = callersExpanded ? faCaretDown : faCaretRight;
+const CallTransportCreator = ({
+  puzzleId,
+  muted,
+  deafened,
+  audioContext,
+  localStream,
+  callersExpanded,
+  onToggleCallersExpanded,
+  selfPeer,
+  otherPeers,
+  router,
+}: {
+  puzzleId: string,
+  muted: boolean;
+  deafened: boolean;
+  audioContext: AudioContext;
+  localStream: MediaStream;
+  callersExpanded: boolean;
+  onToggleCallersExpanded(): void;
+  selfPeer: PeerType;
+  otherPeers: PeerType[];
+  router: RouterType;
+}) => {
+  const [device, setDevice] = useState<types.Device>();
+  useEffect(() => {
+    const newDevice = new Device();
+    newDevice.load({
+      routerRtpCapabilities: JSON.parse(router.rtpCapabilities),
+    }).then(() => setDevice(newDevice));
+  }, [router.rtpCapabilities]);
+
+  useSubscribe(device ? 'mediasoup:transports' : undefined, selfPeer._id, JSON.stringify(device?.rtpCapabilities));
+  const { sendServerParams, recvServerParams } = useTracker(() => {
+    return {
+      // TODO: should we try to detect this specific transport request
+      sendServerParams: Transports.findOne({ peer: selfPeer._id, direction: 'send' }),
+      recvServerParams: Transports.findOne({ peer: selfPeer._id, direction: 'recv' }),
+    };
```

Comment:
@zarvox Interested in your feedback. If, for whatever reason, we churn the `mediasoup:transports` subscription, then potentially we could end up with 2 different `TransportRequest` records and therefore 4 different `Transport` records. In that case, these `findOne`'s could potentially find transports from the older `TransportRequest`, which presumably is about to be torn down. I'm not actually sure when this would happen.

I see two options:

1. We could solve this by having some way to identify the current `TransportRequest`, but there's no good unique ID right now. We'd have to generate a unique ID, thread that through the subscription, and make sure it gets stored onto the `TransportRequest` and copied onto the `Transports`. We'd need to regenerate the ID whenever the subscription got churned, although since I'm not sure when that would happen I don't know when we'd need to churn the ID.
2. We could do nothing and let this theoretical race exist. It should resolve cleanly (eventually the old `Transports` will get deleted and we'll pick up the new ones) so might just create some unnecessary churn

Thoughts?

Reply:
Could always have the findOne sort by createdAt descending if you want to be sure that you’re picking the most up-to-date (to force convergence faster). Otherwise, I wouldn’t worry too much about it unless we’re observing excess churn for whatever strange reason.

Reply:
Yeah I got worried when I realized the race, but having reflected some, I think I'm not actually that worried. I'll edit this comment into a note but I feel OK taking no other action.